core/src/main/scala/caliban/validation/ValueValidator.scala

```diff
@@ -79,14 +79,28 @@ private object ValueValidator {
           case INPUT_OBJECT =>
             argValue match {
               case ObjectValue(fields) =>
-                validateAllDiscard(inputType.allInputFields) { f =>
-                  fields.collectFirst { case (name, fieldValue) if name == f.name => fieldValue } match {
-                    case Some(value)                    =>
-                      validateType(f._type, value, context, s"Field ${f.name} in $errorContext")
-                    case None if f.defaultValue.isEmpty =>
-                      validateType(f._type, NullValue, context, s"Field ${f.name} in $errorContext")
-                    case _                              =>
-                      unit
+                if (inputType._isOneOfInput && fields.size != 1) {
+                  failValidation(
+                    s"$errorContext is not a valid OneOf Input Object",
+                    "OneOf Input Object arguments must specify exactly one non-null key"
+                  )
+                } else if (
+                  inputType._isOneOfInput && fields.size == 1 && fields.headOption.map(_._2).contains(NullValue)
+                ) {
+                  failValidation(
+                    s"$errorContext is not a valid OneOf Input Object",
+                    "OneOf Input Object arguments must specify exactly one non-null key"
+                  )
+                } else {
```

Comment:
Can't we make `Validator.validateOneOfInputValue` package-private (within `validation`) and reuse it here?

Reply:
Done